src/state.rs

```diff
@@ -0,0 +1,219 @@
+// Copyright 2022 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use bls::{PublicKey, PublicKeySet, SecretKey, SecretKeyShare};
+use std::collections::{BTreeMap, BTreeSet};
+use std::sync::Arc;
+
+use crate::error::{Error, Result};
+use crate::knowledge::{Knowledge, KnowledgeFault};
+use crate::sdkg::{AckOutcome, Part, PartOutcome, SyncKeyGen};
+use crate::vote::{DkgSignedVote, DkgVote, IdAck, IdPart, NodeId};
+
+/// State of the Dkg session, contains the sync keygen and currently known Parts and Acks
+/// Can handle votes coming from other participants
+pub struct DkgState<R: bls::rand::RngCore> {
+    id: NodeId,
+    secret_key: SecretKey,
+    pub_keys: BTreeMap<NodeId, PublicKey>,
+    keygen: SyncKeyGen<NodeId>,
+    our_part: Part,
+    all_votes: BTreeSet<DkgSignedVote>,
+    rng: R,
+}
+
+pub enum VoteResponse {
+    WaitingForMoreVotes,
+    BroadcastVote(Box<DkgSignedVote>),
+    RequestAntiEntropy,
+    AntiEntropy(BTreeSet<DkgSignedVote>),
+    DkgComplete(PublicKeySet, SecretKeyShare),
+}
+
+enum DkgCurrentState {
+    IncompatibleVotes,
+    NeedAntiEntropy,
+    Termination(BTreeMap<IdPart, BTreeSet<IdAck>>),
+    WaitingForTotalAgreement,
+    GotAllAcks(BTreeMap<IdPart, BTreeSet<IdAck>>),
+    WaitingForMoreAcks,
+    GotAllParts(BTreeSet<IdPart>),
+    WaitingForMoreParts,
+}
+
+impl<R: bls::rand::RngCore + Clone> DkgState<R> {
+    pub fn from(
+        our_id: NodeId,
+        secret_key: SecretKey,
+        pub_keys: BTreeMap<NodeId, PublicKey>,
+        threshold: usize,
+        rng: &mut R,
+    ) -> Result<Self> {
+        let (sync_key_gen, opt_part) = SyncKeyGen::new(
+            our_id,
+            secret_key.clone(),
+            Arc::new(pub_keys.clone()),
+            threshold,
+            rng,
+        )?;
+        Ok(DkgState {
+            id: our_id,
+            secret_key,
+            pub_keys,
+            keygen: sync_key_gen,
+            all_votes: BTreeSet::new(),
+            our_part: opt_part.ok_or(Error::NotInPubKeySet)?,
```

Comment:
Did you see any reason to have `SyncKeyGen::new()` return an optional Part? I don't really see any reason behind the Observer idea since messages are encrypted between participants, an observer would just be seeing ciphertext flying around.

Reply:
`SyncKeyGen::new()` returns a `None` instead of `Some(part)` when you are an observer but still want a keygen to generate the `PublicKeySet` (but now private key shares). In our use here, all the nodes in `DkgState` are not observers, so we expect `our_part` to be `Some`.

Reply:
Although we don't use them here, observers can also create a `SyncKeyGen` to check all the messages and ultimately generate the `PublicKeySet` from all the `Acks` and `Parts`. They just don't generate a starting `Part` as they are not participants.

Reply:
> messages are encrypted between participants
Messages are partly encrypted but enough data is leaked through so the observers can generate the `PublicKeySet` but not any secret key.